### PR TITLE
[avmfritz] Initial commit of unit tests for model and discovery

### PR DESCRIPTION
**Pull request by cweitkamp**
- Initial commit of unit tests (model and discovery)
- Renamed `JAXBtUtils` to `JAXBUtils`
- Refactoring the use of JAXB Unmarshaller as singleton instance to improve performance

Signed-off-by: Christoph Weitkamp <github@christophweitkamp.de>

### CODE REVIEW

**Comment by cweitkamp** [comment]
@martinvw do you think we get this one merged before finishing #3022? Thanks for short feedback.

File: addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/util/JAXBUtils.java [hunk 1]
```diff
@@ -21,17 +24,31 @@
...
+        }
+    }
+
+    public static DevicelistModel buildResult(final String xml) {
+        try {
+            if (JAXBUnmarshaller == null) {
+                JAXBUnmarshaller = JAXBUtils.JAXBCONTEXT.createUnmarshaller();
+            }
+            return (DevicelistModel) JAXBUnmarshaller.unmarshal(new StringReader(xml));
+        } catch (JAXBException e) {
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/util/JAXBUtils.java]: Why 'handle' it here and return `null` the code should not continue anyway? WDYT?

File: addons/binding/org.openhab.binding.avmfritz.test/src/test/java/org/openhab/binding/avmfritz/internal/ahamodel/AVMFritzModelTest.java [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+                .orElse(null);
+        assertNotNull(device);
+        validateDECT200Model(device);
+
+        device = devices.getDevicelist().stream().filter(it -> "FRITZ!DECT 300".equals(it.getProductName())).findFirst()
+                .orElse(null);
+        assertNotNull(device);
+        validateDECT300Model(device);
+
+        device = devices.getDevicelist().stream().filter(it -> "FRITZ!DECT 301".equals(it.getProductName())).findFirst()
```
[the remark this answers is not in the document]
Comment (cweitkamp) [reply addons/binding/org.openhab.binding.avmfritz.test/src/test/java/org/openhab/binding/avmfritz/internal/ahamodel/AVMFritzModelTest.java]: Yes, why not. I never worked with `java.util.Optional` before, but I will look into that.

File: addons/binding/org.openhab.binding.avmfritz.test/src/test/java/org/openhab/binding/avmfritz/internal/ahamodel/AVMFritzModelTest.java [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+        assertNotNull(device);
+        validateCometDECTModel(device);
+
+        device = devices.getDevicelist().stream().filter(it -> "FRITZ!Powerline 546E".equals(it.getProductName()))
+                .findFirst().orElse(null);
+        assertNotNull(device);
+        validatePowerline546EModel(device);
+    }
+
+    private void validateDECTRepeater100Model(DeviceModel device) {
```
[the remark this answers is not in the document]
Comment (cweitkamp) [reply addons/binding/org.openhab.binding.avmfritz.test/src/test/java/org/openhab/binding/avmfritz/internal/ahamodel/AVMFritzModelTest.java]: Of course.

File: addons/binding/org.openhab.binding.avmfritz.test/src/test/java/org/openhab/binding/avmfritz/internal/ahamodel/AVMFritzModelTest.java [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+        assertEquals(device.getSwitch().getLock(), BigDecimal.ZERO);
+        assertEquals(device.getSwitch().getDevicelock(), BigDecimal.ONE);
+
+        assertEquals(device.getPowermeter().getPower(), new BigDecimal("0.000"));
+        assertEquals(device.getPowermeter().getEnergy(), new BigDecimal("2.087"));
+    }
+
+    @Test
+    public void validateTemperatureConversion() {
+        assertEquals(HeatingModel.fromCelsius(null), BigDecimal.ZERO);
```
[the remark this answers is not in the document]
Comment (cweitkamp) [reply addons/binding/org.openhab.binding.avmfritz.test/src/test/java/org/openhab/binding/avmfritz/internal/ahamodel/AVMFritzModelTest.java]: What a pitty. I din't realize that. Thanks for the hint.

File: addons/binding/org.openhab.binding.avmfritz.test/src/test/java/org/openhab/binding/avmfritz/internal/ahamodel/HeatingModelTest.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+/**
+ * Tests for {@link HeatingModel} methods.
+ *
+ * @author Christoph Weitkamp - Initial contribution
+ *
+ */
+public class HeatingModelTest {
+
+    private final Logger logger = LoggerFactory.getLogger(HeatingModelTest.class);
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.avmfritz.test/src/test/java/org/openhab/binding/avmfritz/internal/ahamodel/HeatingModelTest.java]: Unused?

File: addons/binding/org.openhab.binding.avmfritz.test/src/test/java/org/openhab/binding/avmfritz/internal/discovery/AVMFritzDiscoveryServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+
+/**
+ * Tests for {@link AVMFritzDiscoveryService}.
+ *
+ * @author Christoph Weitkamp - Initial contribution
+ *
+ */
+public class AVMFritzDiscoveryServiceTest {
+
+    private final Logger logger = LoggerFactory.getLogger(AVMFritzDiscoveryServiceTest.class);
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.avmfritz.test/src/test/java/org/openhab/binding/avmfritz/internal/discovery/AVMFritzDiscoveryServiceTest.java]: Unused?

Comment (cweitkamp) [reply addons/binding/org.openhab.binding.avmfritz.test/src/test/java/org/openhab/binding/avmfritz/internal/discovery/AVMFritzDiscoveryServiceTest.java]: Yes, at the moment not in use.

File: addons/binding/org.openhab.binding.avmfritz.test/src/test/java/org/openhab/binding/avmfritz/internal/ahamodel/HeatingModelTest.java (hunk 5 above)
Comment (cweitkamp) [reply addons/binding/org.openhab.binding.avmfritz.test/src/test/java/org/openhab/binding/avmfritz/internal/ahamodel/HeatingModelTest.java]: Yes, at the moment not in use.

**Comment by cweitkamp** [comment]
@martinvw Thanks for review. I changed everything. ;-)

**Pull request merged**